### DISCUSSION TITLE
How would one go about invalidating queries based on a partial argument set ?

### DISCUSSION THREAD

**Discussion by strblr** [n1ru4l/graphql-live-query#616] [Q&A] [answered]
Hello,

Thank you for your awesome work on live queries. I'm strongly considering using it in a large production app.
However, reading the doc, I found myself with the same question I had when I first read about cache invalidation on the client side (for example with `urql`) : How do you invalidate queries without specifying every argument ?

For `urql`'s graphcache, I found a solution [inspired by this](https://spectrum.chat/urql/help/is-there-a-way-to-invalidate-queries-with-variables-without-using-variables~5e4fc1b8-849e-46e9-a9ca-72698a8433d0).

Because not every argument is always relevant. For example if I want to invalidate a paginated `todos` query identified by a `categoryId` (let's say), I wanna target my invalidation only based on `categoryId`, regardless of `cursor`, `limit`, `offset`, etc.

Thank you in advance and keep up the good work.

**Comment by n1ru4l** [comment, marked as answer]
If you have a `Query.todos` field you can invalidate all queries that select the data via the schema coordinate `Query.todos`. Is that what you mean, if not, could you provide a better description and example of what kind of operations you have and how you would like to invalidate them?

**Reply by strblr** [reply]
No that doesn't work because it would invalidate too broadly. I want to be able to target a subset of `Query.todos`, but without supplying all arguments to target them. Is that possible?

**Reply by n1ru4l** [reply]
Okay, I see! I think we could allow this by supplying schema coordinates with argument annotations [`Query.todos(categoryId:)` as described by the schema coordinates RFC](https://github.com/graphql/graphql-spec/blob/main/rfcs/SchemaCoordinates.md#typefieldargname) to the live query store. It can use that information to add those to the tracking "indices", does that sound good?

**Reply by strblr** [reply]
@n1ru4l Does that solve the problem though ? I'll give a concrete example, let's say we have the following schema assuming todos can be classified by category : 

```graphql
type Query {
  todos(categoryId: ID!, page: Int!, limit: Int!): [Todo!]!
}
```

And say we have the following operations from 3 different clients : 

```graphql
query {
  todos(categoryId: "1", page: 2, limit: 5) {
     ...TodoFragment
  }
}
```
```graphql
query {
  todos(categoryId: "1", page: 3, limit: 7) {
     ...TodoFragment
  }
}
```
```graphql
query {
  todos(categoryId: "2", page: 2, limit: 5) {
     ...TodoFragment
  }
}
```

Would `Query.todos(categoryId:"1")` target the two first even though their key signature would be `Query.todos(categoryId:"1",page:2,limit:5)` and `Query.todos(categoryId:"1",page:3,limit:7)` ?

**Reply by n1ru4l** [reply]
Yes, that would totally work. Right now we only index by arguments that are named `id` and are of the `ID!` type ([See this test for an example](https://github.com/n1ru4l/graphql-live-query/blob/977d0d6aaa7929a66a18331c1e16df25c92fb013/packages/in-memory-live-query-store/src/InMemoryLiveQueryStore.spec.ts#L485-L537)). The idea is to configure your live query store on which live query indices it should build for you.

```tsx
new InMemoryLiveQueryStore({
  indexBy: [
    "Query.todos(categoryId:)",
    "Query.user(login:)"
  ]
})
```

Which allows the underlying logic to build indices for those specific topics based on the executed operation

```
Query.todos(categoryId:"1")
Query.todos(categoryId:"2")
Query.user(login:"n1ru4l")
Query.user(login:"dotan")
```

All other args would be dropped from the index name :)

**Reply by strblr** [reply]
Ok, got it. In the big project I was talking about earlier, I have queries that should be indexed on multiple arguments separately. For example, I have `Query.participants` which can be filtered by a `teamId` argument, and by a `sessionId` argument. Sometimes at the same time, sometimes not. So would that mean I have to declare three indexes : 

```javascript
new InMemoryLiveQueryStore({
  indexBy: [
    "Query.participants(teamId:)",
    "Query.participants(sessionId:)",
    "Query.participants(teamId:,sessionId:)"
  ]
})
```
?

If so, I would rather recommend the `graphcache` approach from `urql` : indexing all arguments always, and having an `inspectFields` utility method on the store to retrieve all stored queries on that index. This would remove the need for `indexBy` entirely (which is analogous to Apollo's `@connection` directive), and allow for unlimited custom filtering and targeting of operation indexes.

Like so : 

```javascript
const allFields = store.inspectFields("Query");
const toInvalidate = allFields.filter(field => field.name === "participants" && field.arguments.teamId === "2");
store.invalidate(toInvalidate.map(field => field.index));
```
It would feel annoying after a while to have to keep `indexBy` in sync with all current needs.
Please let me know what your thoughts are.

**Reply by n1ru4l** [reply]
The goal is to have few and light weight indices to keep the backend fast. By complicating the logic there is more overhead and it could potentially slow the whole system down (sequence scan vs hash table/set lookup) and could result in a higher memory footprint. We not only have one client who must be tracked but maybe hundreds. In databases you also have to specify indices explicitly. See how https://github.com/n1ru4l/graphql-live-query/blob/main/packages/in-memory-live-query-store/src/ResourceTracker.ts works.

**Reply by strblr** [reply]
Yes it makes totally sense, but if I understand your code correctly, there might be a way to not only keep the optimized way you're dealing with schema coordinates, but also allow for an option to iterate and custom-filter in very rare cases : That would be to store the arguments as a field of [`StoreRecord`](https://github.com/n1ru4l/graphql-live-query/blob/977d0d6aaa7929a66a18331c1e16df25c92fb013/packages/in-memory-live-query-store/src/InMemoryLiveQueryStore.ts#L255). This way, it would truly feel like a database : indexes for efficiency, but an option to look-up manually in a non-optimized way. 

Since you automatically index every query on `Query.<query name>` regardless of the arguments, one would simply have to call `resourceTracker.getRecordsForIdentifiers("Query.participants")` and might then be able to filter them as desired based on the arguments (which would be a field of the records) and call `run()` when necessary.

**Reply by n1ru4l** [reply]
@ostrebler That might actually work! Would you be interested in implementing such a feature?

**Reply by n1ru4l** [reply]
Another thing I thought about would be just to add an `id` field on the connection, where the id is a composite about the relevant (used) argument values.

```json
{
  "data": {
    "participants": {
      "__typename": "ParticipantsConnection",
      "id": "teamId_1_sessionId_1",
      "edges": []
    }
  }
}
```

```ts
store.invalidate([`ParticipantsConnection:teamId_1_sessionId_1`])
```

Drawbacks? Slightly more data being sent to the client 🤔 and you could argue that the schema now contains live query specific information

**Reply by strblr** [reply]
Yes adding a connection `id` field makes it harder to opt-in / opt-out (especially when you have 25 or so queries) and IMO it exposes too much implementation detail.

I can try and do my best to implement my idea. Sounds like a good challenge!

**Reply by strblr** [reply]
And supplying schema coordinates with custom arguments to `indexBy` would still be very useful as it would allow for full custom indexes. I can look into that, or do you want to do it?

**Reply by n1ru4l** [reply]
@ostrebler I can take care of the index functionality, I actually already started it, but i the meantime I would appreciate your help on working on a API for doing manual lookups!

**Reply by n1ru4l** [reply]
@ostrebler I kicked things off with a small refactor: https://github.com/n1ru4l/graphql-live-query/pull/619 - it should make implementing custom indices quicker. If you need closer contact I am also on the GraphQL Discord, so you can reach me there as well :)

**Reply by strblr** [reply]
@n1ru4l Nice ! I'll start working on my part by the end of the week.

**Reply by strblr** [reply]
Hi @n1ru4l,

My absolute apologies, I got distracted by other tasks.

So here is my situation : I manage a very large Node / GraphQL / React app used by multiple manufacturers and there is now a real call on their part for live updates. So I'm back on where I left in July, but I'm starting to wonder is manual lookup is really that necessary. I think it will depend on what I encounter as I migrate the project to your library. 

Did you implement the coordinate index system? Is it available? 

Last question: some queries produce different output depending on the context and not on the arguments (for example, a todo's "isEditable" field would depend on the privileges of the user making the request). Can that be an issue with your library?

Thank you very much.

**Reply by n1ru4l** [reply]
@ostrebler I have not implemented this yet. All the live queries are tracked individually, thus properties that differ between users will not make a difference! A invalidation will re-execute the whole query operation for each user individually. Of course this might become a bottle neck when having a lot of users. So far I have tested this engine with around 20 users (and very frequent updates) successfully. Potential scaling solutions could be to use more instances so each instances has less operations to re-execute once an invalidation happens. That is something yet to explore and I did not have the right project for that yet.

Please keep in mind that for some application types such as collaborative text editing CRDT might be the better solution.

If you need some reference project you can have a look at https://github.com/dungeon-revealer/dungeon-revealer

**Reply by strblr** [reply]
Hi @n1ru4l, the production app I want to adapt to live queries often has more than 20 users at the same time. I can definitely give you feedback on how it goes.

Using an ugly hack to solve the `mjs` / webpack problem, I was able to kick things off a bit. So far, I tested it with basic queries that have an `id` field, it works like a charm. To go further and since a lot of queries are not indexable by `id`, I actually need the custom coordinate indexing system you talked about [here](https://github.com/n1ru4l/graphql-live-query/discussions/616#discussioncomment-1039026). When may you be able to do it ? Thanks!

I'm starting to work on custom lookups right now. Edit: https://github.com/n1ru4l/graphql-live-query/pull/759
